### PR TITLE
[WIP] GCP component infrastructure

### PR DESCRIPTION
**Pull request by hongye-sun**
The code is for prototype purpose and not in check in state. Please take it as a design review towards gcp core component. Thanks.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/653)

### CODE REVIEW

File: components/gcp/core/kfp/gcp/core/base_op.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
+import signal
```
[the remark this answers is not in the document]
Comment (qimingj) [reply components/gcp/core/kfp/gcp/core/base_op.py]: I think everything under "components" directory are components infrastructures. I would create a new directory for all these reusable components. What about "cloud" directory? "cloud_components"? I don't have a perfect name candidate but would not use the existing components dir.

File: components/gcp/core/kfp/gcp/core/base_op.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+from kubernetes import client, config
+from kubernetes.client.rest import ApiException
+from google.cloud import storage
+import google
+import json
+from datetime import datetime
+import logging
+import sys
+
+class BaseOp:
```
Comment (qimingj) [inline components/gcp/core/kfp/gcp/core/base_op.py]: What I do with comments are:

Any public classes or functions (can be accessed without "_") need to have full docstring (short desc, long desc, args, returns, raises). Any internal classes or functions without a leading "_" should have at least a short desc.

File: components/gcp/dataflow/launch_template_op.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import json
+import logging
+import re
+import time
+
+from googleapiclient import errors
+
+from kfp.gcp.core.base_op import BaseOp
+from kfp.gcp.core import name_utils
```
Comment (qimingj) [inline components/gcp/dataflow/launch_template_op.py]: I don't think you want to expose name_utils. You can do relative imports like https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/compiler/_component_builder.py#L25

File: components/gcp/ml_engine/ml_engine/delete_version_op.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from googleapiclient import errors
+
+from kfp.gcp.core.base_op import BaseOp
+from kfp.gcp.core import name_utils
+from ml_engine import utils
+
+class DeleteVersionOp(BaseOp):
+
+    def __init__(self, project_id, model_name, version_name, staging_dir, wait_interval):
+        super().__init__()
```
Comment (qimingj) [inline components/gcp/ml_engine/ml_engine/delete_version_op.py]: Not super(DeleteVersionOp, self).__init__()?

Comment (Ark-kun) [reply components/gcp/ml_engine/ml_engine/delete_version_op.py]: There is no need for the longer syntax here (and almost everywhere).

File: components/gcp/dataflow/launch_template_op.py (hunk 3 above)
Comment (Ark-kun) [reply components/gcp/dataflow/launch_template_op.py]: >I don't think you want to expose name_utils. You can do relative imports

AFAIK relative imports do not prevent exposure.

File: components/gcp/core/kfp/gcp/core/base_op.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hongye-sun) [reply components/gcp/core/kfp/gcp/core/base_op.py]: Good point. i will address this comment in my breakdown PRs.

File: components/gcp/core/kfp/gcp/core/base_op.py (hunk 2 above)
Comment (hongye-sun) [reply components/gcp/core/kfp/gcp/core/base_op.py]: Will do

File: components/gcp/core/kfp/gcp/core/base_op.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        blog_path = os.path.join(blob_path_prefix, 'staging/{}/{}'.format(
+            self.__class__.__name__,
+            stable_name
+        ))
+        self._storage = storage.Client()
+        self._staging_bucket = staging_bucket
+        self._staging_blob_name = blog_path       
+        self._reload_states()
+
+    def _exit_gracefully(self, signum, frame):
```
[the remark this answers is not in the document]
Comment (hongye-sun) [reply components/gcp/core/kfp/gcp/core/base_op.py]: This is called from SIGTERM signal

File: components/gcp/core/kfp/gcp/core/base_op.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hongye-sun) [reply components/gcp/core/kfp/gcp/core/base_op.py]: A little bit confused about the components folder comment. Isn't the current components folder the root for all reusable components?

**Comment by hongye-sun** [comment]
This PR is a prototype. It's no longer needed.

**Pull request closed without merging**